Sailthru Demo Driver

dylanharbour opened the pull request:
#### Sailthru on 5.3 by Example

This is an example of a basic implementation of Sailthru using Laravel notificatons. This is working, please see screenshots in PR. I've started with a clean install of Laravel 5.3 to illustrate how easy it is to get going. After generating the standard Laravel Auth scaffold I added 1 new notification class called "WelcomeEmailNotification"
. This gets triggered after user registration. In practice, you'd implement one class for each type of notification in the app (ProfileReminderNotification, NewsletterWelcomeNotificaiton etc etc). 

#### Getting Started
* Clone Repo
* Checkout the Sailthru branch (I kept the changes in branch to keep track of easier)
* Composer install, setup your ENV variables / DB / Mail Settings
* Insert your `$apiKey` and `$apiSecret` in the `SailthruChannel.php` (or pull from environment variable/config)
* Register a new user. The `WelcomeEmailNotification` will fire with two notifications. One via normal mail and one via the SailtruChannel


#### Benefits over current implementation

* Lightweight and Laravel friendly. Easy to understand and debug. 

* No need to swap out mail drivers: Sailthru notifications are not treated as emails so they don't interfere with our 
current mail drivers. That makes it a non breaking change to normal mails. 

* Define the Email and Sailthru Messages on the same class. This makes it easier to see what is happening. No need for configs to store various templates / overrides. 

* Run in Parallel. We can run in parallel, or do dynamic switching on the fly between Sailthru and Mail. Making it easier to "warm" the IP / Account. Switching between channels is as easy as modifying the `via()` method on the notification to either go by Mail, Sailthru or both. This can easily be done by a simple global(`config('sailthru.enabled')` type of config, or a more specific one per domain / template / anything else. Switching to a different channel is implemented by design, so it's not complicated.  

* Easy to send Admin mails via mail and other mails via Sailthru.

* Email templates using the mail method are beautifully styled already and super easy to implement / customize / change. 

* 100% fallback. If sailthru goes down, we switch our config off and start sending by normal mail. This *should* be  the case with our current implementation too, but notifications would make it easier to maintain the two channels  together.
 

* Notifications are amazing to test with the Notifications fake. We can use the Notifications fake to test the flow of our our app, and only need to use Mailthief / Mailtrap / Sailthru Fake to test each notification at most, once. Possibly, Developers could do super fast testing while developing with no need to test the external service (Mailthief / Sailthru Test fake).  The last leg to the external service can be run on builds only and can possibly be run in parallel with the rest of  the tests to speed up feedback.  
 
 * Extend past Mail / Sailthru: We can define instant Slack notifications, Push Notifications to our Web Apps, SMS to  phones etc all super easily without having to re-write how / when a notification is triggered. Many major  notifications channels are already supported out of the box like Slack and Push Notification services. 
 
 
#### To Do
*  I did this in just under an hour, it obviously needs more refinement to implement the Client better by a facade / 
 service provider. But that's not too complex if we don't have to hack with our mail driver. 
* Guest users would need a GuestUser class that uses the `notifiable` trait. This class would need to implement the minimum requirements for Sailthru (i.e. Store a Name, Email, etc). We could also make Advertisers Notifiable to send a notification to an Advertiser (i.e. Not to a user - following different rules. ).  
* I've ignored the rest of the Sailthru "features" like Mailing list management.
* Notification `toSailthru()` currently returns an instance of itself. In practice, we'd probably have a sperate  `SailthruNotify` type of  Class which can encapsulate the templat name and variables better.

dylanharbour committed:
Sailthru Demo Driver

* Added New Sailthru Channel which implements the basic send functionality of the Sailthru PHP Client
* New Welcome Email Notification
* Fire Welcome Email Notification after Register Success

dylanharbour commented:
User Signs up: 
<img width="1905" alt="screen shot 2017-06-20 at 12 05 48 am" src="https://user-images.githubusercontent.com/21034071/27320092-df0c83e0-5594-11e7-8fa4-c4223d62602e.png">

Mail Email Fires

<img width="1511" alt="screen shot 2017-06-20 at 12 09 59 am" src="https://user-images.githubusercontent.com/21034071/27320126-06c6222e-5595-11e7-983b-c078b78fe624.png">

Sailthru Email Fires
<img width="1920" alt="screen shot 2017-06-20 at 12 07 14 am" src="https://user-images.githubusercontent.com/21034071/27320137-15702004-5595-11e7-9652-25ccb934549f.png">

Logged in Sailthru Correctly
<img width="1917" alt="screen shot 2017-06-20 at 12 06 39 am" src="https://user-images.githubusercontent.com/21034071/27320156-23fa9a78-5595-11e7-8531-27e7d3ca33df.png">

sperelson commented on app/Notifications/WelcomeEmailNotification.php:
```diff
@@ -0,0 +1,91 @@
+use Illuminate\Contracts\Queue\ShouldQueue;
+use Illuminate\Notifications\Messages\MailMessage;
+
+class WelcomeEmailNotification extends Notification
```
We would need to make them queueable
` implements ShouldQueue`

https://laravel.com/docs/5.3/notifications#queueing-notifications

dylanharbour replied:
Correct. Super easy to make them queueable.

sperelson commented on app/Notifications/WelcomeEmailNotification.php:
```diff
@@ -0,0 +1,91 @@
+        return [
+            'mail',
+            SailthruChannel::class
+        ];
```
Magic would happen here. Due to how each venture chooses the correct channel per email and per domain, we would have special logic here to determine the correct (and only) driver based on some value sent with the `$notifiable`.

dylanharbour replied:
I would make it explicit here since each notification is defined in a known context. 

So this could be something like 

```
if (domain_config('sailthru.enabled')) {
   return SailthruChannel::class;
} else {
   return ['mail'];
} 


```

If this was an admin notification like `Admin\ReportAbuseNotification`, we'd just hardcode it to go  by mail,
```
return ['mail']

```
 or use one single config
```
return domain_config('mail.admin_driver')
```
or via multiple channels like 
```
return [domain_config('mail.admin_driver'), 'slack']

```

By being explicit within the notification, we can drop the need for complicated configs and magic.

dylanharbour replied:
We could also make Each of the notifications extend a class in which the via() method is implemented, like `WelcomeEmailNotification` extends `AbstractCustomerNotification`. 

We could have `AbstractAdminNotification` `AbstractAdvertiserNotification` `AbstractGuestNotification` etc

sperelson replied:
When queuing the notification, at what point does the `via()` method get executed? Before being queued or after?

sperelson replied:
Also, Cheki makes this more complicated as we configure the email driver per email and per domain.

dylanharbour committed:
Added Queueing

*  Implement Database Queue to show when each method is being invoked

e.g. User Registered with QUEUE not running.
[2017-06-28 08:09:55] local.DEBUG: In RegisterController registered Method
[2017-06-28 08:09:55] local.DEBUG: In WelcomeEmailNotification Construct
[2017-06-28 08:09:55] local.DEBUG: In WelcomeEmailNotification Via Method

Then queue started with `php artisan queue:work`
[2017-06-28 08:10:20] local.DEBUG: In WelcomeEmailNotification toMail Method
[2017-06-28 08:10:24] local.DEBUG: In WelcomeEmailNotification toSailthru Method

dylanharbour commented:
@sperelson I've changed this to a queued notification to illustrate when each method gets fired. 
https://github.com/dylanharbour/sailthru-demo/pull/1/commits/dd177444be30fe931f5d067738c1d0f7b572a9fe

In Summary, the `construct` and `via` methods are fired before putting the Notification on the queue. The actual `toMail` or `toSailthru` is fired when the queue is worked. 

```
e.g. User Registered with QUEUE not running.
[2017-06-28 08:09:55] local.DEBUG: In RegisterController registered Method
[2017-06-28 08:09:55] local.DEBUG: In WelcomeEmailNotification Construct
[2017-06-28 08:09:55] local.DEBUG: In WelcomeEmailNotification Via Method

Then queue started with `php artisan queue:work`
[2017-06-28 08:10:20] local.DEBUG: In WelcomeEmailNotification toMail Method
[2017-06-28 08:10:24] local.DEBUG: In WelcomeEmailNotification toSailthru Method
```
